text/0000-let-expression.md

```diff
@@ -0,0 +1,917 @@
+    && eq_expr_value(cx, lhs0, rhs1)
+    && eq_expr_value(cx, lhs1, rhs0)
+    || continue;
+    // 30 lines of code with two less tab
+}
+```
+Every `if let` or `if_chain!` that fill body of a loop or function can refactored in this way. You can easily find
+dozens of them just in rust-clippy. Note that if-let-chain alone can't solve this problem, because you have access to
+variables of if-let inside of block, but you have access to variables of a binding statement under it, thus you can
+save one indentation.
+
+`||` is not only useful for let-else style things. This is a real example from [deno](https://github.com/denoland/deno):
+
+```rust
+let nread = if let Some(s) = resource.downcast_rc::<ChildStdoutResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<ChildStderrResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<TcpStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<TlsStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<UnixStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.read(buf).await?
+} else {
+    return Err(not_supported());
+};
+```
+Which by `||` counterpart of if-let-chain can become:
+```rust
+let nread = if let Some(s) = resource.downcast_rc::<ChildStdoutResource>()
+    || let Some(s) = resource.downcast_rc::<ChildStderrResource>()
+    || let Some(s) = resource.downcast_rc::<TcpStreamResource>()
+    || let Some(s) = resource.downcast_rc::<TlsStreamResource>()
+    || let Some(s) = resource.downcast_rc::<UnixStreamResource>()
+    || let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.read(buf).await?
+} else {
+    return Err(not_supported());
+};
+```
+This is smaller and doesn't repeat a code.
+
+A different class of practical usages of this RFC is let expression usage as a bool. People
+wrap their let expressions with `if expr { true } else { false }` manually. This need is almost met
+with `matches!` macro, but `if let true else false` is still a thing in rust code bases. Again from rust-clippy:
+```rust
+fn is_repeat_zero(&self, expr: &Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::Call(fn_expr, [repeat_arg]) = expr.kind;
+        if is_expr_path_def_path(self.cx, fn_expr, &paths::ITER_REPEAT);
+        if let ExprKind::Lit(ref lit) = repeat_arg.kind;
+        if let LitKind::Int(0, _) = lit.node;
+
+        then {
+            true
+        } else {
+            false
+        }
+    }
+}
+```
+After this RFC, we can write this:
+```rust
+fn is_repeat_zero(&self, expr: &Expr<'_>) -> bool {
+    let ExprKind::Call(fn_expr, [repeat_arg]) = expr.kind
+    && is_expr_path_def_path(self.cx, fn_expr, &paths::ITER_REPEAT)
+    && let ExprKind::Lit(ref lit) = repeat_arg.kind
+    && matches!(lit.node, LitKind::Int(0, _)) // you can use let expression here as well
+}
+```
```

Comment:
Is this better just because it is shorter and without extra indentation?

You need strong arguments for why this should be a language feature.

Reply:
Indentation isn't important here, point is `if` with branches equal to `true` or `false` are often reinventing bool operations (`if b { true } else { false }` is reinvention of nothing operation) and those operations read better than their `if` equivalent (I added a better example).

> You need strong arguments for why this should be a language feature.

Arguments are together strong. This is not the only argument for this RFC.